pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+func NewCassandraBlobAccess(capabilitiesProvider capabilities.Provider, readBufferFactory blobstore.ReadBufferFactory, tlsConfig *tls.Config, port, protoVersion int32, username, password string, clusterHosts []string, keyspace, tablePrefix string, segmentSizeInBytes int32, preferredDc string, lastAccessUpdateInterval time.Duration, universalInstanceName string) blobstore.BlobAccess {
+	cassandraBlobStorePrometheusMetrics.Do(func() {
+		prometheus.MustRegister(segmentWriteHist)
+		prometheus.MustRegister(segmentReadHist)
+		prometheus.MustRegister(totalWritesHist)
+		prometheus.MustRegister(lastAccessUpdatesFailures)
+	})
+
+	cluster := gocql.NewCluster(clusterHosts...)
+	cluster.Keyspace = keyspace
+	cluster.Consistency = writeConsistency
+	if preferredDc != "" {
+		cluster.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(gocql.DCAwareRoundRobinPolicy(preferredDc))
+	}
+
+	if port != 0 {
+		cluster.Port = int(port)
+	}
+
+	if protoVersion != 0 {
+		cluster.ProtoVersion = int(protoVersion)
+	}
+
+	if username != "" {
+		cluster.Authenticator = gocql.PasswordAuthenticator{
+			Username: username,
+			Password: password,
+		}
+	}
+
+	if tlsConfig != nil {
+		cluster.SslOpts = &gocql.SslOptions{
+			Config: tlsConfig,
+		}
+	} else {
+		log.Printf("Not setting up TLS config")
+	}
+
+	if tablePrefix == "" {
+		log.Printf("No table prefix set. Cowardly refusing to proceed.")
+		return blobstore.NewErrorBlobAccess(errors.New("table prefix must be set"))
+	}
+	log.Printf("Table prefix: %s", tablePrefix)
+
+	log.Printf("Creating session in keyspace '%s' for %v", keyspace, clusterHosts)
+
+	session, err := cluster.CreateSession()
+	if err != nil {
+		log.Printf("Unable to create session: %v", err)
+		return blobstore.NewErrorBlobAccess(err)
+	}
+
+	// Create the worker pool for updating the last access times
+	jobs := make(chan job, lastAccessUpdateWorkerCount)
+	for i := 0; i < lastAccessUpdateWorkerCount; i++ {
+		go lastUpdateWorker(jobs)
+	}
+
+	log.Printf("Cassandra storage ready.")
+
+	return &cassandraBlobAccess{
+		capabilitiesProvider:     capabilitiesProvider,
+		session:                  session,
+		readBufferFactory:        readBufferFactory,
+		segmentSize:              segmentSizeInBytes,
+		lastAccessUpdateInterval: lastAccessUpdateInterval,
+		lastAccessUpdate:         jobs,
+		universalInstanceName:    universalInstanceName,
+		tables:                   newTables(tablePrefix, session),
+	}
+}
+
+func (a *cassandraBlobAccess) GetCapabilities(ctx context.Context, instanceName bbdigest.InstanceName) (*remoteexecution.ServerCapabilities, error) {
+	return a.capabilitiesProvider.GetCapabilities(ctx, instanceName)
+}
+
+func logErrorIfNotCancelledContext(err error, msg string, args ...interface{}) {
+	if err != nil && !errors.Is(err, context.Canceled) {
+		initialMessage := fmt.Sprintf(msg, args...)
+		log.Printf("%s: %v", initialMessage, err)
+	}
+}
```

Comment:
Please remove logging like this. If we just bubble up errors properly, then all of the existing tracing middleware should provide sufficient insight. If not, we should look into setting up https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/github.com/gocql/gocql.